### PR TITLE
rvar and draws_rvars interfaces

### PR DESCRIPTION
**Pull request by mjskay**
Alright, I finally got around to finishing a first version of this while procrastinating other things :).

This has implementations and mostly complete (I think) tests for the `rvar` and `draws_rvars` data types. I apologize for the giant bolus of code, and I am sorry if it takes awhile to review.

For the `draws_rvars()` format I tried to implement as much of the API as is present in other formats as I could, though I very easily may have missed something. Please let me know if I did not get the expected behavior correct here or there. 

Some notes:

- In the end I decided to leave chain information in the `rvar()` format and just suffer through making it work, as I had already partially implemented it using an attached attribute that stores the number of chains anyway. One consequence is that `repair_draws()` for the `draws_rvars` format will try to detect if draws have been re-ordered, in which case it assumes chain information is no longer valid and drops the chains.

- `bind_draws.draws_rvars()` currently works for all values of `along` except "iteration". In principle I'm sure one could do binding for that format along iterations but the implementation seemed not immediately pleasant so I didn't get around to it.

- At some point (probably not too soon though :) ) I'd like to add support for factor-type rvars as they would be useful for samples from discrete distributions, but haven't gotten to that yet. I don't think this would be too hard (basically would implement `levels()` and a few other factor-related functions).

- If there's interest I could throw together a vignette on the `rvar` format.

- I tried to keep new dependencies to a minimum, but this does add a couple so I thought I should provide justification:
  - vctrs - needed for helper functions for implementing the `rvar()` vector type. This is an upstream dependency of tibble, which posterior already depends on.
  - tensorA - used for implementing matrix multiplication of rvars as a tensor product
  - pillar - used for applying styled text in rvar output (particularly graying out the "+/-sd" when shown in tibbles). This is an upstream dependency of tibble, which posterior already depends on.
  - distributional - used to import generics (currently `cdf()`) for distribution vectors like rvar for compatibility with the distributional package. In the future I may also provide support for casting to/from the distributional data type.

Happy to take suggestions for reducing the number of deps if that is a concern.

### CODE REVIEW

**Comment by paul-buerkner** [comment]
Thank you so much @mjskay!! This PR looks quite impressive indeed! I will take a closer look and run some tests in detail next week.

**Comment by jgabry** [comment]
Awesome, thanks @mjskay! I will also try to go through this next week if possible.

**Comment by mjskay** [comment]
I did a few more API cleanup things this week and added a vignette to introduce the basics of the rvar datatype, on the assumption that might be helpful to have for review.

Two other questions came to mind as I did that:

- I'm currently a little inconsistent in some function names about prefixing some things (like `nchains`) with `.` or not. In some cases it's probably useful to avoid potential name collisions with `...` arguments, in others it isn't but I used it anyway (mostly because you always had a roxygen doc template for `.nchains`). I'm now thinking I should perhaps change it to `nchains` everything except where `...` is used (and possibly change some other args to have a `.` prefix in those cases). Let me know if that sounds reasonable. I always struggle with these dot-prefix things, debating whether it's best to be consistent in arg names across functions or to consistently apply the "only use `.` when there are `...` args that could take arbitrary names" rule.

- I'm not very happy with the name of the `rvar_r()` function and would love suggestions for a better name. There's an example of it in the vignette.

**Comment by paul-buerkner** [comment]
I am starting to review your PR in more depth and I must admist I am impressed by it. Nice work!

With regard to your latest comments: 

- I agree with you that we should be using `.` only in places where collision with `...` should be avoided.
- I will check `rvar_r` and think of a better name.

File: tests/testthat/test-extract_variable.R
Comment (paul-buerkner) [inline tests/testthat/test-extract_variable.R:22]: do not forget to add a test once the bug I noted for `extract_variable.draws_rvars` has been fixed.

**Review by paul-buerkner — CHANGES_REQUESTED** [review]
I have revied the complete PR except for the vignette and some of the vctrs interface (because I lack experience with vctrs) and I am really impressed the interface you created! Thank you so much! I did some minor cleaning directly in the rv-like branch and have added a couple of (mostly minor) comments in my review.

I feel like `rvar`s could be the main output objects for a lot of post-processing function in brms etc. going forward, although this is perhaps a separate discussion.

**Comment by mjskay** [comment]
Thanks @paul-buerkner for the kind words and thorough review! I am swamped in teaching/service stuff for the next week at least so it might be a bit before I get through your comments, but I will try to do so soon.

File: R/rvar-rfun.R
Comment (mjskay) [reply R/rvar-rfun.R:12]: Yeah, I've used that style in the past and have come to dislike it slightly. I find it cumbersome to write APIs against, because if you wrap a function `f` with that has a parameter `x` defined by an expression with a function `g` that has a different default for `x`, if your user wants to make a call to `g` that uses the default of `x` from `f` they have to recreate the entire expression instead of passing a simpler value. 

But this doesn't bother me that much in this case (that usage would be rare here). I'll switch to the other style if you prefer it.

File: R/as_draws_matrix.R
Comment (mjskay) [reply R/as_draws_matrix.R]: Sure, done. I had to make `enclos` an argument of `do_call` in order to support its usage in `rdo`. I'll open a separate comment for that in case you want a different solution.

File: R/rvar-rfun.R
Comment (mjskay) [inline R/rvar-rfun.R:167]: I had to add an `enclos` argument to `do_call` so that I could replace `do.call` with `do_call` here. Let me know if you'd prefer a different solution.

File: R/bind_draws.R
Comment (mjskay) [reply R/bind_draws.R:183]: Sounds good :). Removing this TODO.

File: R/print.R
Comment (mjskay) [reply R/print.R:305]: Hah nope, that must have been a typo. Removing.

File: R/misc.R
Comment (mjskay) [reply R/misc.R:258]: Sure, done.

File: R/rvar-.R
Comment (mjskay) [reply R/rvar-.R:1]: It's a convention I copied from some other packages (I think ggplot uses it?). It ensures `rvar-.R` comes before the other `rvar-*.R` files in directory sort order (`rvar.R` would come after them). I don't think it makes a big difference though (I don't think there's anything in the rvar files that currently depends on collation order), so if you'd prefer it without the `-` I will remove it here and in the test file name. Let me know.

File: R/rvar-.R
Comment (mjskay) [reply R/rvar-.R]: done.

File: R/rvar-.R
Comment (mjskay) [reply R/rvar-.R:574]: I had copied this style after [this stackoverflow issue](https://stackoverflow.com/a/39621289) and [usage in tidyverse packages](https://github.com/tidyverse/dplyr/blob/master/R/reexport-tibble.r), which generates a `reexport.Rd` file with links to re-exported functions. But I tried what you suggested and it seems to work better (it doesn't generate the redundant reexports.Rd file but everything else --- links etc --- seems to still work). So I've switched to that. Perhaps that style was not intended for use with S3 generics...

File: R/rvar-.R
Comment (mjskay) [reply R/rvar-.R:557]: Sure --- `density.rvar` seems the natural place. Did you have specific language in mind?

File: R/rvar-.R
Comment (mjskay) [reply R/rvar-.R:93]: I admit it feels a bit weird. Basically I experimented with several different formats, and this was the format that I was able to get to work with {vctrs} and the various {tidyverse} functions (e.g. so that you could include rvars in a tibble and functions like `tidyr::pivot_wider()` would still work). Other obvious candidates were not great for various reasons (a subclass of arrays didn't work properly with some tidyverse functions for reasons I couldn't quite figure out --- I think its length was measured incorrectly internally because some code bypassed the length() function I had overriden; an S4 class appeared to have other problems integrating with tibbles). The format I came down to still isn't perfect, and I think it in future it could be worth exploring different approach (maybe something with ALTREP?), but this format was the best I could do to get something working with the widest range of functions.

File: R/rvar-math.R
Comment (mjskay) [reply R/rvar-math.R:1]: Good point, I renamed it `rvar-summaries`

File: R/rvar-math.R
Comment (mjskay) [reply R/rvar-math.R:79]: Thanks! :)

File: R/rvar-math.R
Comment (mjskay) [reply R/rvar-math.R:121]: good idea, done.

File: R/rvar-.R
Comment (mjskay) [reply R/rvar-.R:574]: nevermind --- I tried the method you suggested and I got warnings about undocumented code objects, which perhaps explains why the approach I was using is recommended. I have reverted to the approach I used about and I am unresolving this conversation in case there is something I misunderstood about your suggestion.

File: R/rvar-rfun.R
Comment (mjskay) [reply R/rvar-rfun.R:222]: Ah yeah, that's nice! Changing.

File: R/rvar-.R
Comment (mjskay) [reply R/rvar-.R:61]: No, it shouldn't cause any problems to defer it. 

At some point it would probably be good to set up proper casting to/from the {distributional} data types, but I don't think it's necessary for the first version. I'll open a separate issue and remove the TODO from here.

File: R/rvar-.R
Comment (mjskay) [reply R/rvar-.R:491]: Should we find someone more familiar with it? It being so new (and kind of convoluted...) I'm not sure who it would be.

File: R/rvar-math.R
Comment (mjskay) [reply R/rvar-math.R:124]: Summary is a "group generic" like `Math` and `Ops` (see `help("groupGeneric")`) used to implement a set of generic functions all at once (`all`, `any`, `sum`, `prod`, etc). I agree the resulting help topic name is confusing though... let me change it to `rvar-summaries-by-draw` as a parallel to `rvar-summaries` (which is what I changed the `E` topic to based on another comment).

File: R/rvar-math.R
Comment (mjskay) [reply R/rvar-math.R:212]: ah yeah, missed that. done.

File: R/rvar-math.R
Comment (mjskay) [reply R/rvar-math.R:317]: From profiling, the time in this function seems to be dominated by `tensorA::chol.tensor()` at the moment. Looking at it, `chol.tensor()` seems to be implemented using an lapply so that's not too surprising (unlike `tensorA::mul.tensor()` which I used to implement matrix multiplication, which calls down to a C implementation).

So I'd say the bigger problem is `chol.tensor()`... possibly we need to find a better tensor library and/or implement it ourselves if we want them to be performant? It looks like other functions that might make sense to have (`solve`, `svd`, ...?) will have similar issues.

If you don't object, I'd say we start by getting the initial implementation in on this PR, then can replace them with faster implementations later? Likely someone with more knowledge than I have of tensor libraries in R would be the right person to do that.

File: R/rvar-print.R
Comment (mjskay) [reply R/rvar-print.R:67]: Fair. Part of my motivation here is that on the off chance we might want to change the structure later, I think it is good to hide it a bit. Generally speaking people shouldn't need to access any internals other than the backing array anyway, which `draws_of()` gives them.

File: R/rvar-print.R
Comment (mjskay) [reply R/rvar-print.R:28]: I originally went with the condensed form so that things would chunk better in vector and matrix output, but I see your point. I added the spaces around +/- and made sure there were 2 spaces between elements in vector/matrix/tibble output. I think that works better than what I had before.

File: R/extract_variable.R
Comment (mjskay) [reply R/extract_variable.R]: Good catch, I added a check for this and a more helpful error message.

File: R/extract_variable_matrix.R
Comment (mjskay) [reply R/extract_variable_matrix.R:42]: Added a more helpful error message here too.

File: tests/testthat/test-extract_variable.R
Comment (mjskay) [reply tests/testthat/test-extract_variable.R:22]: done

File: R/rvar-.R
Comment (mjskay) [reply R/rvar-.R:2]: Good point. I split things up into what is hopefully a more logical set of subfiles. Let me know if you think things should be organized differently now.

File: R/rvar-.R
Comment (mjskay) [reply R/rvar-.R]: (per the comment on splitting out this file, this stuff has been moved to rvar-cast.R where you made a similar comment, so I am resolving this one here)

File: R/rvar-cast.R
Comment (mjskay) [reply R/rvar-cast.R:1]: Should we find someone more familiar with it? It being so new (and kind of convoluted...) I'm not sure who it would be.

File: R/rvar-math.R
Comment (mjskay) [reply R/rvar-math.R]: I've implemented it now.

**Comment by mjskay** [comment]
@paul-buerkner thanks again for the detailed review!

I've gone through everything and either addressed it (resolving the corresponding comment) or left comments/questions for you. Mostly minor stuff; I think I resolved most of the bigger things.

File: R/rvar-rfun.R
Comment (paul-buerkner) [reply R/rvar-rfun.R:12]: I see. Both options are fine by me. Feel free to leave it as is.

File: R/rvar-math.R
Comment (paul-buerkner) [reply R/rvar-math.R:323]: I agree to your suggestion.

File: R/rvar-rfun.R
Comment (paul-buerkner) [reply R/rvar-rfun.R:167]: Looks good to me. I have had way to many problems with do.call in brms, especially if two or more do.calls happend to be nested in one-another, so I wanted to make sure, we do not run into the same trouble here.

File: R/rvar-print.R
Comment (paul-buerkner) [reply R/rvar-print.R:67]: I agree that it is good to hide as long as we state somewhere explicitely that we aim to hide the latent structure as much as possible and recommend the `draws_of()` function to extract it if required for some reason.

**Comment by paul-buerkner** [comment]
Thank you for checking and responding to my comments. I am quite happy with this PR in its present form and have no further objections. With regard to the reviewing of the `vctrs` stuff, I am ok with just merging even if no expert has reviewed this part. Before merging, I would like to also get an approval of @jgabry. You don't have to fully review this PR but I wouldn't want to merge it without your approval.

**Comment by paul-buerkner** [comment]
@jgabry Do you have time to approve this PR as well? Again, no need to check it in detail since this is quite a huge PR.

**Review by paul-buerkner — APPROVED** [review]

**Comment by mjskay** [comment]
FYI @paul-buerkner I just made some minor changes since you reviewed this PR (see https://github.com/stan-dev/posterior/pull/106/commits/f5ca7e863bd3e64858ae4d1b3450d5600a598469). These changes make `rvar` indexing a bit more similar to base array indexing. I added support for matrix-based indexing (e.g. `x[rbind(c(1,2,3), c(4,5,6))]`, which is roughly equivalent to `c(x[1,2,3], x[4,5,6])`). I also changed indexing with `x[i]` on a multidimensional array to work as in base arrays; i.e. it indexes to the `i`th element in a flattened form of the array. Previously in `rvar`s this would return the `i`th row, adding however many trailing empty indices are needed to do that --- that behavior can still be achieved with `rvar`s by using a single comma (e.g. `x[i,]`).

I made these changes because both forms of indexing are needed to properly support the base `diag()` function---which seemed like something folks might want---and I couldn't think of any negative consequences.

**Comment by paul-buerkner** [comment]
Thanks! Those changes sound good to me!

**Comment by paul-buerkner** [comment]
Even as Jonah is not responding, I would like to move this PR forward anyway. Is there anything else you want to change before we merge?

**Comment by mjskay** [comment]
I've been trying it out by implementing some things on top of it in tidybayes, which led me to iron out a few other corners, so I'm more confident it's a decent first version. I'd love to see it merged and get more testing and feedback!

**Comment by mjskay** [comment]
(wait --- just need to fix the change I just made that I stupidly thought would be simple but wasn't)

**Comment by mjskay** [comment]
okay, should be good now.

**Comment by paul-buerkner** [comment]
Merging this now. Thank you so much for providing this exciting feature!

**Comment by mjskay** [comment]
My pleasure! Thanks for the help and feedback!

**Comment by jgabry** [comment]
Just wanted to say that this is super cool. Sorry I didn't help review this (I finally now have some time to work on posterior). And @mjskay thanks a ton for implementing this!

**Comment by mjskay** [comment]
> Just wanted to say that this is super cool. Sorry I didn't help review this (I finally now have some time to work on posterior). And @mjskay thanks a ton for implementing this!

Thanks @jgabry, it's been my pleasure!

**Pull request merged**
